agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/PythonRefAgentConnection.kt

```diff
@@ -57,137 +97,242 @@ class PythonRefAgentConnection : AgentConnection {
+                                 *
+                                 * If a caller is locked in the loop for longer time, it's going to be interrupted after timeout.
+                                 */
+                                Thread.sleep(1000)
+                            }
+                        }
+                    }
+                }
+            } catch (e: Throwable) {
+                observer.onError(e)
+            }
         }
     }
-
-    fun waitForMessageOfType(type: String): String {
-        var messageOfType: String? = null
-        while (messageOfType == null) {
-            sleep(500)
-            messageOfType = popMessageOfType(type)
+    /**
+     * Waits for incoming connection from whatever party which possesses the invite
+     */
+    override fun waitForInvitedParty(invite: String): Single<IndyPartyConnection> {
+        return Single.create { observer ->
+            try {
+                if (getConnectionStatus() == AgentConnectionStatus.AGENT_CONNECTED) {
+                    /**
+                     * On an incoming connection request, the agent must send the "request_received"
+                     */
+                    webSocket.receiveMessageOfType<RequestReceivedMessage>(MESSAGE_TYPES.REQUEST_RECEIVED).subscribe({
+                        /**
+                         * On the "request_received" message, reply with "send_response" with remote DID to any incoming connection request.
+                         * At this stage it doesn't matter which party sent the connection request, because it's not possible to correlate
+                         * "request_received" message and the invite. For the purpose of PoC we reply with "send_response"
+                         * on each "request_received", no matter which party is on the other side.
+                         *
+                         * TODO: suggest an improvement in pythonic indy-agent that incorporates the invite's public key in the "request" message,
+                         * TODO: so that it's possible to correlate "request_received" which includes "request" and the invite
+                         */
+                        sendAsJson(RequestSendResponseMessage(it.did))
+                        /**
+                         * Wait until the STATE has pairwise connection corresponding to the invite.
+                         */
+                        val pubKey = getPubkeyFromInvite(invite)
+                        pollStatusForPairwiseConnection(pubKey).timeout(60000, TimeUnit.MILLISECONDS).subscribe({ pairwise ->
+                            /**
+                             * Wait for "response_sent" from DID (their_did) associated with the invite public key
+                             */
+                            webSocket.receiveMessageOfType<RequestResponseSentMessage>(MESSAGE_TYPES.RESPONSE_SENT, pairwise["their_did"].asText()).subscribe { responseSent ->
```

Comment:
I move this logic away to generateInvite function, see above

Reply:
fixed